Lenin del rio react project week2b

Lenindelrionicaragua opened the pull request:
I've completed almost all the requirements, but there are a few CSS details I couldn't replicate. One of them is changing the color of the link once it has been clicked, and the other is centering the product description text vertically with respect to the image. This time it was much more challenging to get everything working. I'm sure there are several things that can be improved. I look forward to your advice.

tlorent commented on week2/project/src/App.css:
```diff
@@ -0,0 +1,24 @@
+.product-list {
+  display: grid;
+  gap: 20px;
+  grid-auto-flow: row; /* This allows items to flow into a new row when there's not enough space in the column */
+}
+
+/* Responsive styles */
+@media (min-width: 576px) {
+  .product-list {
+    grid-template-columns: repeat(2, 1fr);
+  }
+}
+
+@media (min-width: 768px) {
+  .product-list {
+    grid-template-columns: repeat(3, 1fr);
+  }
+}
+
+@media (min-width: 992px) {
+  .product-list {
+    grid-template-columns: repeat(4, 1fr);
+  }
+}
```
The columns are looking good now! 🙌 But there is actually a way to do this all in one line with `grid-template-columns` and you don't have to use the media queries 🙂 Also, for the media queries you don't have to add `.product-list` all the time, you also do the media queries inside the `.product-list` class.

```suggestion
.product-list {
  display: grid;
  gap: 20px;
  grid-template-columns: repeat(auto-fill,minmax(300px,1fr));
}
```

tlorent commented on week2/project/src/App.js:
```diff
@@ -0,0 +1,79 @@
+  }, []);
+
+  const handleCategoryClick = (category) => {
+    // Handle category click
```
These comments are not necessary, the name of the function itself already explains what it does 🙂 

```suggestion
```

tlorent commented on week2/project/src/App.js:
```diff
@@ -0,0 +1,79 @@
+    setSelectedProduct(null);
+  };
+
+  const renderLoadingOrError = () => {
+    // Render loading or error message
+    if (loading) {
+      return <p>Loading...</p>;
+    } else if (error) {
+      return <p>{error}</p>;
+    }
+  };
```
Nice job adding the loading and error displaying in a separate place, not in `return` to make it more readable! 👍 

I would detach them even further to be honest, separate it even more and because they're not entirely related to each other. Also, I would not necessarily create a separate function for it because the final `return` should be responsible for rendering the main content.

```suggestion
    if (loading) {
      return <p>Loading...</p>
    }
    
    if (error) {
      return <p>{error}</p>
    }
```

tlorent commented on week2/project/src/App.js:
```diff
@@ -0,0 +1,79 @@
+  };
+
+  const renderHeader = () => {
+    // Render header with title and application routes
```
```suggestion
```

tlorent commented on week2/project/src/App.js:
```diff
@@ -0,0 +1,79 @@
+
+  const renderHeader = () => {
+    // Render header with title and application routes
+    return (
+      <React.Fragment>
+        <h1>{selectedProduct ? selectedProduct.title : "Products"}</h1>
+        <AppRoutes
+          categories={categories}
+          selectedCategory={selectedCategory}
+          setSelectedProduct={setSelectedProduct}
+          products={allProducts}
+          handleCategoryClick={handleCategoryClick}
+        />
+      </React.Fragment>
+    );
+  };
+
+  return (
+    <Router>
+      <div className="App">
+        <header className="App-header">
+          {renderLoadingOrError() || renderHeader()}
+        </header>
+      </div>
+    </Router>
+  );
+};
```
I would just add this in the main `return` 🙂 Also, you can use `<></div>`, the shorthand for a fragment 👍 

```suggestion
  return (
    <Router>
      <div className="App">
        <header className="App-header">
          <>
            <h1>{selectedProduct ? selectedProduct.title : "Products"}</h1>
            <AppRoutes
              categories={categories}
              selectedCategory={selectedCategory}
              setSelectedProduct={setSelectedProduct}
              products={allProducts}
              handleCategoryClick={handleCategoryClick}
            />
          </>
        </header>
      </div>
    </Router>
  );
};
```

tlorent commented on week2/project/src/AppRoutes.js:
```diff
@@ -0,0 +1,66 @@
+    <Route
+      path="/"
+      element={
+        <React.Fragment>
```
You can use the shorthand here, we don't really write `React.Fragment` anymore unless you need to apply styling to it via a class or need to add other attributes (like a `key` when mapping)

```suggestion
        <>
```

tlorent commented on week2/project/src/AppRoutes.js:
```diff
@@ -0,0 +1,66 @@
+      path="/"
+      element={
+        <React.Fragment>
+          {/* Component for displaying a list of categories */}
+          <CategoryList
+            categories={categories}
+            selectedCategory={selectedCategory}
+            onCategoryClick={handleCategoryClick}
+          />
+
+          {/* Component for displaying a list of products */}
```
The comments are not necessary because the components already explain what happens here. Which means you write clear code and as another developer and reader of your code I clearly understand what is going on 👍 

```suggestion
          <CategoryList
            categories={categories}
            selectedCategory={selectedCategory}
            onCategoryClick={handleCategoryClick}
          />
```

tlorent commented on week2/project/src/components/CategoryList/CategoryList.js:
```diff
@@ -0,0 +1,24 @@
+    <ul className="category-list">
+      {categories.map((category) => (
+        <CategoryListItem
+          key={category}
```
The `category` value is not really unique among siblings, right? 🙂 All electronics have the same category, all jewelry have the same category. Try to find another value you can use 😉

Lenindelrionicaragua replied:
I corrected the use of category.id as the key

tlorent commented on week2/project/src/components/ProductList/ProductList.js:
```diff
@@ -0,0 +1,70 @@
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState(null);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const productsData = await getProducts(selectedCategory);
+        setProducts(productsData);
+        setLoading(false);
+      } catch (error) {
+        setError("Error fetching products");
+        setLoading(false);
+      }
+    };
+
+    fetchData();
+  }, [selectedCategory]);
```
🙌

tlorent commented on week2/project/src/components/ProductList/ProductList.js:
```diff
@@ -0,0 +1,70 @@
+  }, [selectedCategory]);
+
+  const handleProductClick = (product) => {
+    // Ensure that onProductClick is a function before invoking it
+    if (typeof onProductClick === "function") {
+      onProductClick(product);
+    }
+  };
```
Can you explain why this is necessary? 🙂 Does it now work without this check?

Lenindelrionicaragua replied:
By removing the precaution of checking whether it is a function or not, the code threw an error indicating that it was not a function. In an attempt to address this, I considered that the issue might be in how I exported the function from the ProductList component, but attempting to fix it did not yield positive results.

To tackle this situation, I decided to define the function in the App.js file and then passed it as a property through Routes to the ProductList file. However, I am still puzzled by the fact that the code worked correctly when including the if statement as a precaution. In summary, I made the aforementioned changes, and now the code operates without the need for this precaution anywhere in it.

I appreciate your guidance to better understand this situation.

tlorent reviewed (commented):
@Lenindelrionicaragua Nice work! 🙌 Very impressive the code you are already writing. I have some suggestions/comments, please have a look at them when you can 😄

Lenindelrionicaragua commented:
> @Lenindelrionicaragua Nice work! 🙌 Very impressive the code you are already writing. I have some suggestions/comments, please have a look at them when you can 😄

Thank you very much for the advice on handling ProductList and optimizing the CSS; it is much more concise now.

I made some corrections:

I removed the renderLoadingOrError function and will render the loading and error messages directly within the return statement in the App.js file.
I eliminated the use of React.Fragment.
I removed unnecessary comments.
In CategoryList, I corrected the use of category.id as the key.
The most significant issue was addressing the handling of the handleOnClickProduct function. To resolve this, I moved the function to the main App.js file and exported it to ProductList. I appreciate you pointing this out, as it has given me a better understanding of function invocation times.

Once again, thank you for your valuable guidance."
